src/chapters/4-property-driven-training.md

```diff
@@ -229,48 +229,98 @@ The definition of the optimisation problem above instantiates by taking:
         \mathcal{L}_{\mathcal{S}(\mathbf{x})} = \mathcal{I} ( || f(\mathbf{x}) - f(\hat{\mathbf{x}})|| \leq \delta)
     $$
 
-# Coding Example: generating a logical loss functions for mnist-robustness in Vehicle
+# Coding Example: generating a logical loss function in Python
 
-To generate a logical loss function usable in Python, we will use Vehicle's Python bindings that come pre-installed.
+The Python bindings now ship backend-specific helpers in `vehicle_lang.loss.tensorflow`
+and `vehicle_lang.loss.pytorch`. They both expose a single entry point
+`load_specification(path, *, logic=..., samplers=..., declarations=..., declaration_context=...)`
+which compiles your `.vcl` file to a dictionary of callable Python loss functions.
 
-We open a new Python file and write:
+- `logic` defaults to `DifferentiableLogic.DL2`, but you can also request `DifferentiableLogic.Vehicle`.
+- `samplers` lets you override the search strategy for Vehicle `search` expressions.
+  If omitted, a default FGSM-based sampler is used (see `DefaultTensorFlowSampler` and
+  `DefaultPyTorchSampler`).
+- The returned dictionary keys are the names of `@property` declarations in your spec.
+
+Below is a minimal PyTorch example that mirrors the tests in `vehicle-python/tests`:
 
 ```python
-from vehicle_lang.compile import Target, to_python
+import torch
+from vehicle_lang.typing import DifferentiableLogic
+from vehicle_lang.loss.pytorch import load_specification
+
+# Compile the Vehicle spec to PyTorch loss functions
+spec = load_specification(
+    "test_trainable.vcl",  # any Vehicle spec path
+    logic=DifferentiableLogic.Vehicle,
+)
 
-spec = to_python(
-    "mnist-robustness.vcl",
-    target=Target.LOSS_DL2,
-    samplers={"pertubation": sampler_for_pertubation},
+constraint_loss = spec["output_bounded"]  # name of a @property in the spec
+model = torch.nn.Sequential(
+    torch.nn.Linear(1, 8), torch.nn.ReLU(), torch.nn.Linear(8, 1)
 )
+optimizer = torch.optim.Adam(model.parameters(), lr=1e-2)
 
-robust_loss_fn = spec["robust"]
-```
+def network(x):
+    return model(x.reshape(1, 1)).reshape(1)
+
+for step in range(50):
+    optimizer.zero_grad()
 
-which generates the loss function representing the logical loss function for the property `robust`.
+    # Task loss (e.g., MSE to a target function)
+    task_loss = torch.tensor(0.0)
+    for x_val, y_val in [(0.0, 0.0), (0.5, 1.0), (1.0, 2.0)]:
```

Comment:
I think it would be easier to understand if you split this out into a top-level declaration called `dataset` which you pre-cast to a `torch.tensor`? Same for the Tensorflow example below.